docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+
+Note that type capabilities, unions and intersections are types themselves.
+So `(A & imm)` is completely different from `(A & mut)` or `(A & iso)`.
+Specifically, you cannot convert `mut` to `imm` (as you would with a `const` cast in C++), or vice versa.
+You can freeze an `iso` region into `imm`, but unfreeze an `imm` region.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific non-trivial runtime semantics.
+
+The type system guarantees that:
+ * Structural soundness: Interfaces and classes form valid concrete types, operations and functions take in the correct (sometimes inferred) types.
+ * Capability guarantees are followed: The compiler checks operations on types, the runtime guarantees the semantics is followed.
+ * Runtime safety: A program that compiles is a program that runs safely, without crashes, traps or unhandled exceptions, including asynchronous behaviour, but not necessarily producing the right result or doing what you expect.
+
+## Classes and Interfaces
+
+Classes and interfaces have a similar structure:
+ * Fields: elements of a specific type that have a runtime space allocated.
+ * Methods: functions that belong to a type.
+
+The declaration is trivial:
+```ts
+// Helps receivers know they can convert the object to a string representation
+interface Serialisable
+{
+  // All derived classes will have the common field
+  var common : F64;
+
+  // Derived classes must implement this method
+  toString() : String;
+};
+
+class MyType : Serialisable
+{
+  // Inherits `common`
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `toString`
+  toString() : String
+  {
+    ...
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later.
+
+They can also be a private member, if its name starts with `_`.
+
+They can also have a keyword `embed`, which means ...
+
+### Methods
+
+Methods are functions that belong to a class or interface. Like functions, they can have arguments and a return type.
+
+Methods can be called statically (via `Class::method()`) or dynamically (via `object.method()`).
+On the static case, the `self` pointer is a descriptor forthe type, while on the dynamic case, it's a descriptor to the instance (and its vtable, type info, etc).
+
+Static calls cannot change the fields of the class unless an object is being created and returned, like in the case of constructors.
+
+### Constructors and Destructors
+
+Verona distinguishes memory allocation with object construction.
+There are no `constructors` per se, in Verona, only static methods that can allocate memory (using `new`), build an object and return it.
```

Comment:
I would phrase that as something more like:
```suggestion
Verona's built-in `new` function fully constructs an object and so must be called with values for all of the fields in the object.
It may be used only from static methods on the class and so is usually used by factory methods.
By convention, factory methods are called `create()` and there is syntactic sugar to allow `T()` to implicitly call `T::create()`.
```